typeclass-experiments/src/Common/Types.hs

```diff
@@ -0,0 +1,175 @@
+
+       The final Pat should be instantiated to a Pat body.
+
+       In some languages, parts of this may be ignored. E.g. in Rust the type name doesn't matter (we use the constr name of the
+       outermost inner sum for constructing types). -}
+    DecP Pat Pat Pat
+  deriving (Show, Eq, Ord)
+
+infixr 5 :*
+
+-- Pattern synonyms. These VASTLY improve readability
+
+pattern List :: Pat -> Pat
+pattern List t = AppP ListP t
+
+pattern Maybe :: Pat -> Pat
+pattern Maybe t = AppP MaybeP t
+
+pattern Map :: Pat -> Pat -> Pat
+pattern Map k v = AppP (AppP MapP k) v
+
+pattern Either :: Pat -> Pat -> Pat
+pattern Either l r = AppP (AppP EitherP l) r
+
+pattern RecFields :: Pat -> Pat -> Pat -> Pat
+pattern RecFields l x xs = RecP ((l := x) :* xs)
+
+pattern ProdArgs :: Pat -> Pat -> Pat
+pattern ProdArgs x xs = ProdP (x :* xs)
+
+pattern RecBody :: Pat -> Pat -> Pat
+pattern RecBody constr fields = SumP (constr := RecP fields :* Nil)
+
+pattern ProdBody :: Pat -> Pat -> Pat
+pattern ProdBody constr fields = SumP ((constr := ProdP fields) :* Nil)
+
+pattern Rec :: Pat -> Pat -> Pat -> Pat -> Pat
+pattern Rec tyName tyVars constr fields =
+  DecP tyName tyVars (SumP (constr := RecP fields :* Nil))
+
+pattern Product :: Pat -> Pat -> Pat -> Pat -> Pat
+pattern Product tyName tyVars constr args =
+  DecP tyName tyVars (SumP (constr := ProdP args :* Nil))
+
+pattern SumBody :: Pat -> Pat -> Pat -> Pat
+pattern SumBody l x xs = SumP ((l := x) :* xs)
+
+pattern Sum :: Pat -> Pat -> Pat -> Pat
+pattern Sum tyName tyVars pat =
+  DecP tyName tyVars pat
+
+pattern Void :: Pat
+pattern Void = SumP Nil
+
+{- Utility functions. Turn a list of types into a product/record/sum type.
+-}
+toProd :: [Pat] -> Pat
+toProd = ProdP . foldr (:*) Nil
+
+toRec :: [Pat] -> Pat
+toRec = RecP . foldr (:*) Nil
+
+toSum :: [Pat] -> Pat
+toSum = SumP . foldr (:*) Nil
+
+{- Utility functions. Turn a list of types into a product/record/sum type.
+-}
+defToPat :: Ty.TyDef -> Pat
+defToPat Ty.TyDef {..} = DecP (Name tyDefName) vars $ case tyDefBody of
+  Ty.Sum constrs -> toSum . NE.toList . fmap goConstr $ constrs
+  where
+    collectFreeTyVars (Ty.TKind xs) = foldr (\x acc -> Name x :* acc) Nil xs
+
+    vars = collectFreeTyVars tyDefKind
+
+    goConstr :: (Ty.ConstrName, Ty.Product) -> Pat
+    goConstr (n, p) = Name n := goProduct p
+
+    goProduct :: Ty.Product -> Pat
+    goProduct = \case
+      Ty.Empty -> Nil
+      Ty.Record rMap -> toRec . NE.toList . fmap (uncurry (:=) . bimap Name goPat) $ rMap
+      Ty.Product pList -> toProd . NE.toList . fmap goPat $ pList
+
+    goPat :: Ty.SourceTy -> Pat
+    goPat = \case
+      Ty.TyVar t -> VarP t
+      Ty.TApp a b -> AppP (goPat a) (goPat b)
+      Ty.PrimT prim -> goPrim prim
+      Ty.TyRef ref -> case ref of
+        Ty.Local t -> RefP (Name t)
+        _ -> undefined -- dunno what to do here. Imports should be resolved by this stage?
```

Comment:
Oh I misunderstood how foreign refs work I think. You're saying that foreign refs are refs to other *LambdaBuffs* modules and not to other modules in the target language, right? 

Isn't it a big problem that we don't allow users to import types from other modules in the target languages? That seems like a *much* bigger practical limitation than e.g. the lack of recursive types would be. 

I guess they could define them as opaque types... but if we let them *use* those types it would break our ability to derive typeclass instances unless the users submitted a PR implementing instance rules for them. 

As annoying as ps-bridge was, the fact that you could specify "translate Haskell type `X` into PS type `Y` from module `Z`" was borderline essential for our use of it, was it not?

Reply:
That's what `Opaque` types are, and as you suggested, you have to implement type-def mappings and whatever `instance` was declared for it. That's essentially the same, but devoid from ps-bridge hackery.

This is very much a 'LambdaBuffers FFI' feature, any and all primitive/builtin/opaques are treated the same way; if you declare an opaque, you must manually implement codegen aspects of it, namely type defs and typeclass semantics.

We should aim at the following Opaques:
1. Platform agnostic standard library, perhaps even getting feature parity with Protobufs (int, bytes, composites like arrays/maps/sets)
2. Plutus standard library that declares the entirety of the `plutus-ledger-api` types as Opaques

As we implement these, we will hopefully converge to an 'easy to use' codegen framework that is also robust (testing).

Reply:
We do not attack the use case of `I already have an application running with its own types, now I want to use LB` and take the route like others in this space where any shared types must be declared in LB from the very start.